include/dali/core/convert.h

```diff
@@ -78,165 +79,312 @@ struct ret_type {  // a placeholder for return type
+#endif
+  }
+
+  DALI_HOST_DEV
+  static constexpr Out ConvertSatNorm(In value) {
+#ifdef __CUDA_ARCH__
+    return std::is_signed<Out>::value
+      ? clamp<Out>(detail::cuda_round_helper(value * max_value<Out>(), Out()))
+      : detail::cuda_round_helper(max_value<Out>() * __saturatef(value), Out());
+#else
+    return clamp<Out>(std::round(value * max_value<Out>()));
+#endif
+  }
+};
 
-template <typename Out>
-constexpr __device__ __host__ std::enable_if_t<std::is_unsigned<Out>::value, Out>
-ConvertSatNorm(float value) {
+/// Converts signed to signed, unsigned to unsigned or unsigned to signed
+template <typename Out, typename In,
+          bool IsOutSigned = std::is_signed<Out>::value,
+          bool IsInSigned = std::is_signed<In>::value>
+struct ConvertIntInt {
+  DALI_HOST_DEV
+  static constexpr Out Convert(In value) { return value; }
+  DALI_HOST_DEV
+  static constexpr Out ConvertNorm(In value) {
+    return Converter<Out, float>::Convert(value * (1.0f * max_value<Out>() / max_value<In>()));
+  }
+  DALI_HOST_DEV
+  static constexpr Out ConvertSat(In value) { return clamp<Out>(value); }
+  DALI_HOST_DEV
+  static constexpr Out ConvertSatNorm(In value) {
+    return ConvertNorm(value);
+  }
+};
+
+/// Converts signed to unsigned integer
+template <typename Out, typename In>
+struct ConvertIntInt<Out, In, false, true> {
+  DALI_HOST_DEV
+  static constexpr Out Convert(In value) { return value; }
+  DALI_HOST_DEV
+  static constexpr Out ConvertNorm(In value) {
+    return Converter<Out, float>::Convert(value * (1.0f * max_value<Out>() / max_value<In>()));
+  }
+  DALI_HOST_DEV
+  static constexpr Out ConvertSat(In value) { return clamp<Out>(value); }
+  DALI_HOST_DEV
+  static constexpr Out ConvertSatNorm(In value) {
 #ifdef __CUDA_ARCH__
-  return max_value<Out>() * __saturatef(value);
+    return detail::cuda_round_helper(
+      __saturatef(value * (1.0f / max_value<In>())) * max_value<Out>());
 #else
-  return max_value<Out>() * (value < 0.0f ? 0.0f : value > 1.0f ? 1.0f : value);
+    return value < 0 ? 0 : ConvertNorm(value);
 #endif
+  }
+};
+
+/// Converts between integral types
+template <typename Out, typename In>
+struct ConverterBase<Out, In, false, false> : ConvertIntInt<Out, In> {
+};
+
+/// Pass-through conversion
+template <typename T>
+struct Converter<T, T> {
+  static DALI_HOST_DEV
+  constexpr T Convert(T value) { return value; }
+
+  static DALI_HOST_DEV
+  constexpr T ConvertSat(T value) { return value; }
+
+    static DALI_HOST_DEV
+  constexpr T ConvertNorm(T value) { return value; }
+
+  static DALI_HOST_DEV
+  constexpr T ConvertSatNorm(T value) { return value; }
+};
+
+template <typename raw_out, typename raw_in>
+using converter_t = Converter<
+  std::remove_cv_t<raw_out>,
+  std::remove_cv_t<std::remove_reference_t<raw_in>>>;;
+
+}  // namespace detail
+
+template <typename Out, typename In>
+DALI_HOST_DEV constexpr Out Convert(In value) {
+  return detail::converter_t<Out, In>::Convert(value);
```

Comment:
Maybe you could provide short doc for these 4 functions, describing what precisely conversion is performed here? Cause it's not obvious reading the code of this functions ;)

Reply:
Done.